CondFormats/SiStripObjects/src/SiStripApvSimulationParameters.cc

```diff
@@ -0,0 +1,80 @@
+#include "CondFormats/SiStripObjects/interface/SiStripApvSimulationParameters.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "CLHEP/Random/RandFlat.h"
+
+namespace {
+  PhysicsTools::Calibration::HistogramF2D calculateZInt(const SiStripApvSimulationParameters::LayerParameters& params) {
+    auto hZInt = PhysicsTools::Calibration::HistogramF2D(params.upperLimitsX(), params.upperLimitsY());
+    for (int i{0}; i != params.numberOfBinsX() + 2; ++i) {
+      for (int j{0}; j != params.numberOfBinsY() + 2; ++j) {
+        float zInt = 0.;
+        for (int k{0}; k != params.numberOfBinsZ() + 2; ++k) {
+          zInt += params.binContent(i, j, k);
+        }
+        hZInt.setBinContent(i, j, zInt);
+      }
+    }
+    return hZInt;
+  }
+}  // namespace
+
+void SiStripApvSimulationParameters::calculateIntegrals() {
+  if (m_barrelParam.size() != m_barrelParam_zInt.size()) {
+    m_barrelParam_zInt.resize(m_barrelParam.size());
+    for (unsigned int i{0}; i != m_barrelParam.size(); ++i) {
+      m_barrelParam_zInt[i] = calculateZInt(m_barrelParam[i]);
+    }
+  }
+}
+
+bool SiStripApvSimulationParameters::putTIB(SiStripApvSimulationParameters::layerid layer,
+                                            SiStripApvSimulationParameters::LayerParameters&& params) {
+  if ((layer > m_nTIB) || (layer < 1)) {
+    edm::LogError("SiStripApvSimulationParameters")
+        << "[" << __PRETTY_FUNCTION__ << "] layer index " << layer << " out of range [1," << m_nTIB << "]";
+    return false;
+  }
+  m_barrelParam[layer - 1] = params;
+  m_barrelParam_zInt[layer - 1] = calculateZInt(params);
+  return true;
+}
+
+bool SiStripApvSimulationParameters::putTOB(SiStripApvSimulationParameters::layerid layer,
+                                            SiStripApvSimulationParameters::LayerParameters&& params) {
+  if ((layer > m_nTIB + m_nTOB) || (layer < m_nTIB + 1)) {
+    edm::LogError("SiStripApvSimulationParameters") << "[" << __PRETTY_FUNCTION__ << "] layer index " << layer
+                                                    << " out of range [" << m_nTIB + 1 << "," << m_nTOB << ")";
+    return false;
+  }
+  m_barrelParam[m_nTIB + layer - 1] = params;
+  m_barrelParam_zInt[m_nTIB + layer - 1] = calculateZInt(params);
+  return true;
+}
+
+float SiStripApvSimulationParameters::sampleBarrel(layerid layerIdx,
+                                                   float z,
+                                                   float pu,
+                                                   CLHEP::HepRandomEngine* engine) const {
+  if (m_barrelParam.size() != m_barrelParam_zInt.size()) {
+    throw cms::Exception("LogicError") << "z-integrals of 3D histograms have not been calculated";
+  }
+  const auto layerParam = m_barrelParam[layerIdx];
+  const int iz = layerParam.findBinX(z);
+  const int ip = layerParam.findBinX(pu);
```

Comment:
Just caught my eye as I was skimming through - should one of these be findBinY()?

Reply:
Thanks, fixed